operators/pkg/controller/elasticsearch/driver/downscale.go

```diff
@@ -14,122 +15,221 @@ import (
+}
+
+// attemptDownscale attempts to decrement the number of replicas of the given StatefulSet,
+// or deletes the StatefulSet entirely if it should not contain any replica.
+// Nodes whose data migration is not over will not be removed.
+// A boolean is returned to indicate if a requeue should be scheduled if the entire downscale could not be performed.
+func attemptDownscale(ctx downscaleContext, downscale ssetDownscale, allLeavingNodes []string, statefulSets sset.StatefulSetList) (bool, error) {
+	// TODO: only one master node downscale at a time
+	if downscale.canRemoveStatefulSet() {
+		ssetLogger(downscale.statefulSet).Info("Deleting statefulset")
+		if err := ctx.k8sClient.Delete(&downscale.statefulSet); err != nil {
+			return false, err
+		}
+		return false, nil
 	}
 
-	// TODO: don't remove last master/last data nodes?
-	// TODO: detect cases where data migration cannot happen since no nodes to host shards?
+	if downscale.targetReplicas >= downscale.initialReplicas {
+		// nothing to do
+		return false, nil
+	}
 
-	for _, node := range leavingNodes {
-		if migration.IsMigratingData(observedState, node, leavingNodes) {
-			// data migration not over yet: schedule a requeue
-			logger.V(1).Info("Data migration not over yet, skipping node deletion", "node", node)
-			reconcileState.UpdateElasticsearchMigrating(resourcesState, observedState)
-			results.WithResult(defaultRequeue)
+	// adjust the theoretical downscale to one we can safely perform
+	performable, shouldRequeue := calculatePerformableDownscale(ctx, downscale, allLeavingNodes)
+
+	// do the performable downscale
+	return shouldRequeue, doDownscale(ctx, performable, statefulSets)
+}
+
+// calculatePerformableDownscale updates the given downscale target replicas to account for nodes
+// which cannot be safely deleted yet.
+// It returns the updated downscale and a boolean indicating whether a requeue should be done.
+func calculatePerformableDownscale(
+	ctx downscaleContext,
+	downscale ssetDownscale,
+	allLeavingNodes []string,
+) (ssetDownscale, bool) {
+	// create another downscale based on the provided one, for which we'll slowly decrease target replicas
+	performableDownscale := ssetDownscale{
+		statefulSet:     downscale.statefulSet,
+		initialReplicas: downscale.initialReplicas,
+		targetReplicas:  downscale.initialReplicas, // target set to initial
+	}
+	// iterate on all leaving nodes (ordered by highest ordinal first)
+	for _, node := range downscale.leavingNodeNames() {
+		if migration.IsMigratingData(ctx.observedState, node, allLeavingNodes) {
+			ssetLogger(downscale.statefulSet).V(1).Info("Data migration not over yet, skipping node deletion", "node", node)
+			ctx.reconcileState.UpdateElasticsearchMigrating(ctx.resourcesState, ctx.observedState)
 			// no need to check other nodes since we remove them in order and this one isn't ready anyway
-			break
+			// make sure we requeue
+			return performableDownscale, true
```

Comment:
I take that back, other ssets are not blocked and for a single sset it seems there is no way around it. Should we have any recommendation about multiple ssets (partitioning) for large clusters? Any changes would be only applied after slowest node is ready.

Reply:
Yes, this is per StatefulSet. We stop downscaling one StatefulSet but will still downscale the others if possible.

> Slowest node would be blocking the removal of half of the nodes (avg. case).
I don't think there's an easy way around that with StatefulSets. We **have to** delete the node with highest ordinal first. I think we can generally expect the same performance for nodes in the same StatefulSet (same storage, config and resources).

Reply:
Easy way probably not, but we could mitigate it by recommending (or managing on our own) multiple ssets per sset defined by user. Apart from some negative implications it would achieve less surprising behavior (instead of wondering why x nodes are waiting for this single one taking minutes/hours to move out from). I guess it's kind of going back to not having ssets:) I think it's safe to ignore this until (if?) we will hear about issues around it.